CMake: (fix) Autotools build HDF5

jwsblokland opened the pull request:
- Fixed the issue for not able to find the hdf5.mod Fortran module file
  in the case case that HDF5 has been built using the autotools.
  This has been reported in the following issue: https://github.com/LLNL/H5Z-ZFP/issues/57
- Fixed incorrectly set of the variable H5Z_ZFP_WITH_OPENMP in
  the h5z_zfp-config.cmake file.

jwsblokland commented:
As you can read, this pull request is addressing autotools build HDF5 version mentioned in issue https://github.com/LLNL/H5Z-ZFP/issues/57. Somehow one of the Fortran tests fails while I have not changed any of the source files. I ran the tests myself without any problem. So I am a bit list what the problem could be.

jwsblokland commented:
I found the problem. It turns out that for certain set of random numbers one of the Fortran tests fails. The generated random number are based on a seed generation by teh processor. I switched to a fixed seed, such the set of random numbers will always be the same and that the tests do not fail.

brtnfld commented:
I changed the tolerances for the Fortran test in PR 58, so this should fix the Fortran failures.

markcmiller86 commented on test/test_rw_fortran.F90:
```diff
@@ -73,34 +74,34 @@ END FUNCTION real_eq
      CALL GET_COMMAND_ARGUMENT(i,arg,len,status)
      IF (status .NE. 0) THEN
-        WRITE (*,*) 'get_command_argument failed: status = ', status, ' arg = ', i
+        WRITE (ERROR_UNIT,*) 'get_command_argument failed: status = ', status, ' arg = ', i
```
@brtnfld are you ok with changes to output error messages from `WRITE (*,*)` here to `WRITE (ERROR_UNIT, *)`? In particular, take note that messages for when things pass (lines 291/294) are written to `OUTPUT_UNIT`.

brtnfld replied:
I'm ok with that.

markcmiller86 commented on test/test_rw_fortran.F90:
```diff
@@ -305,27 +306,34 @@ END PROGRAM main
   REAL(C_DOUBLE) :: rand
-  pi = 4.0_C_DOUBLE*ATAN(1.0_C_DOUBLE)
 
+  CALL RANDOM_SEED(SIZE=size)
```
@jwsblokland I completely agree that using a *fixed* seed is better.

Please pardon my Fortran ignorance (its been many years since I've coded much in Fortran) but I am wondering why having a fixed seed then requires `ALLOCATABLE`? Why not something like `integer, parameter :: seed = 123456789` in the function variable declaration block?

For that matter, @brtnfld now that I am here thinking about it, would you object to `real, parameter :: pi = 3.14159265358979` instead of what you are doing with basically a *derived* formula?

brtnfld replied:
The formula adjusts the maximum precision tailored to the architecture but making it a constant also works provided you have enough digits for the precision (and some extra for peace of mind).

I also think it is better to not have a fixed seed since I would rather test for possible extremes and account for them in the test.

markcmiller86 replied:
> The formula adjusts the maximum precision tailored to the architecture but making it a constant also works provided you have enough digits for the precision (and some extra for peace of mind).

Ok, so we're talking at most double precision here, right? So, 16 digits is plenty for that, right? [This ref](http://www.mimirgames.com/articles/programming/digits-of-pi-needed-for-floating-point-numbers/) offers some of the constant values. If we had quad precision, we'd need ~33 digits. This 40 digit value is probably fine... `3.1415926535897932384626433832795028841971`

> I also think it is better to not have a fixed seed since I would rather test for possible extremes and account for them in the test.

I am not sure I follow. If we believe the test is a good test regardless of seed, I think it should be repeatable and so use a fixed seed. What *more* does not having a fixed seed do for us?

brtnfld replied:
Oddly, the test fails for certain values; to me, the test should not depend on the random numbers generated whether it fails or passes.  But I think that was because the tolerance in the values check was too extreme. I think this has been addressed, so it should be ok to use a fixed seed. The Fortran test is mainly for testing the wrappers anyway.

jwsblokland replied:
@markcmiller86 Regarding the allocatable. This is how the subroutine random_seed() has been implemented (see [gfortran random_seed](https://gcc.gnu.org/onlinedocs/gfortran/RANDOM_005fSEED.html) for more details). The be honest, I was also expected one seed value.

markcmiller86 commented:
I wanted to create a stacked PR with a proposed change to this PR. But, to do that, I think I need write access to your fork. So, I am proposing a change here in comments. It removes *allocatable*-related stuff and makes both `SEED` and `PI` fixed parameters. I don't know if the code compiles though as I have no easy access to fortran. Lemme know if this looks like a decent direction. It certainly simplifies things a bit IMHO.

```
! Generate a simple, 1D sinusioidal data array with some noise
SUBROUTINE gen_data(npoints, noise, amp, buf)
  USE ISO_C_BINDING
  IMPLICIT NONE
  
  INTEGER, PARAMETER :: SEED=123456789
  REAL, PARAMETER :: PI = 3.1415926535897932384626433832795028841971

  INTEGER(C_SIZE_T) :: npoints
  REAL(C_DOUBLE) :: noise
  REAL(C_DOUBLE) :: amp
  REAL(C_DOUBLE), DIMENSION(1:npoints) :: buf

  INTEGER(C_SIZE_T) :: i
  REAL(C_DOUBLE) :: x
  REAL(C_DOUBLE) :: pi
  REAL(C_DOUBLE) :: rand

  ! Initializatiions
  CALL RANDOM_SEED(PUT=SEED)
  pi = PI
  
  DO i = 1, npoints
     rand = REAL(i, C_DOUBLE)
     CALL RANDOM_NUMBER(rand)
     x = 2_c_double * pi * REAL(i-1, C_DOUBLE) / REAL(npoints-1, C_DOUBLE)
     buf(i) = amp*( 1.0_C_DOUBLE + SIN(x)) + (rand - 0.5_C_DOUBLE)*noise
  ENDDO
```

jwsblokland commented:
I rebased this pull request to the latest version of the `master` branch. Furthermore, I have implemented proposed suggestions:
- introduction of quad precision PI value
- cast this value to a double one which will be used in the computation

brtnfld commented on test/test_rw_fortran.F90:
```diff
@@ -375,27 +376,38 @@ END PROGRAM main
   REAL(C_DOUBLE) :: amp
   REAL(C_DOUBLE), DIMENSION(1:npoints) :: buf
 
+  REAL(REAL128), PARAMETER :: PI = 3.1415926535897932384626433832795028841971_REAL128
```
What if the platform does not support 128? Does not this cause an error? What benefit is this over just making it a double and over-prescribe the precision?

jwsblokland replied:
That is a valid point. According to the 2008 Fortran Standard the REAL128 will have a negative value if the target platform does not support it. Let me change it to REAL64, assuming that this is defined for all platforms. This also means that we can only specify PI up to 12 digits after the period.

brtnfld replied:
I would just make it C_DOUBLE. Fortran does not complain about extra digits, it will just truncate the number.

jwsblokland replied:
That is a good idea.

brtnfld reviewed (approved):
LGTM

markcmiller86 commented:
Looks like I was misunderstanding what `seed` was here. Its actually an *array* of values, not necessarily a single, scalar number. We could opt to do something similar for it as in...

```
     integer, dimension(20) :: seed

     seed = (/ 0,1,3,5,7,11,13,17,19,23,27,29,31,33,37,41,43,47,53,57 /)

     call random_seed(put=seed)
```

But, we'd be guessing whether `20` was big enough for all implementations. 

Also, I tested `random_seed()` without arguments and for me, on gcc anyways, it produces repeatable results. Nevertheless, I like the idea of coding things here to ensure reproducibility.

Thanks @jwsblokland for the fix 😉
